### PR TITLE
Add better GC for concurrent map and fix several memory leaks as suggested by Valgrind

### PR DESCRIPTION
**Pull request by ikavalio**
Hi Richard,

This PR adds fixes to 2 issues:
1. Improved GC algorithm for concurrent map. Previous algorithm was simple reference counting, that modified shared state too frequently (on each call to get/put/delete). Current algorithm is epoch based quiescent states memory reclaimer inspired by [QSBR](http://preshing.com/20160726/using-quiescent-states-to-reclaim-memory/). It uses mutexes, but that's fine, since reader thread can enter quiescent state once per N iterations.
2. There were a couple of places (signal handler destructor, unit tests) where Valgrind has found memory leaks.

### CODE REVIEW

File: src/main/cpp/concurrent_map.h [hunk 1]
```diff
@@ -460,9 +455,9 @@ struct Migration : public JobCoordinator::Job {
 				m->prev = this;
 
 				origTable->coordinator.set(m);
-        	} else {
-        		TRACE(LFMap, 21);
-        	}
+			} else {
```
Comment (RichardWarburton) [inline src/main/cpp/concurrent_map.h]: formatting looks a bit weird here.

**Comment by RichardWarburton** [comment]
Hi @ikavalio, thanks for the PR. Maybe I've misunderstood things, but we can't use mutexes anywhere within a signal handler, as they aren't re-entrant safe functions. So I don't think I can accept all these changes if they add mutex use to the signal handler.

**Comment by ikavalio** [comment]
Hi Richard,

Maybe I'm missing something, but I don't think map is used within signal handlers... All inserts/deletes happen during `ThreadStart` / `ThreadEnd` events while signals are blocked for current thread. All gets are done in `LogWriter::record` that is called from `Processor::run` by processing thread (SIGPROF signals are also blocked). Please correct me if I'm wrong...

**Comment by RichardWarburton** [comment]
Thanks for clarifying @johnoliver

**Pull request merged**